### ISSUE TITLE
Recent releases do not include a sdist

### ISSUE THREAD

**Issue by jayvdb** [tox-dev/tox-travis#120]
https://pypi.org/simple/tox-travis/

The last sdist was 0.8, with latest release at 0.11

**Comment by ryanhiebert** [comment]
I did that on purpose, as I didn't see a need to have multiple variants to install on one platform. Is there a reason that you think we should make both formats?

**Comment by jayvdb** [comment]
sdists include all the necessary files to install and test the software, and e.g. atm I am wanting one in order to create an RPM of tox-travis, specifically because it is a mandatory dependency at https://github.com/ssato/python-anytemplate/issues/7 .  While it shouldnt be a mandatory dependency, these things happen, so having tox-travis available as an RPM reduces a hurdle that other users may face.

**Comment by jayvdb** [comment]
Packaged at https://build.opensuse.org/package/show/home:jayvdb:moban/python-tox-travis

**Comment by ryanhiebert** [comment]
I'm really not keen to release an sdist. The case you're mentioning is obviously a bug in the upstream software. I don't think that it should be a bug in this package that someone else is depending on this when they really shouldn't be.

Trying to follow the links, the issue that you pulled up is about not including tests, not about removing this dependency. Is there another issue that you were intending for me to look at?

If it turns out that there's more than a few packages like this and it would really be helpful to release an sdist, I'll be willing to reconsider. I just think it's really icky, so there's going to have to be several packages that I'd be helping before that sounds okay to me.

Is it reasonable for you to apply a patch to fix the dependency issue before you package this software? How are other bugs dealt with downstream?

**Comment by rpkilby** [comment]
> sdists include all the necessary files to install and test the software

It's worth mentioning that currently, the sdist does not contain the tests anyway.

<details>
<summary>python setup.py sdist</summary>

```
running sdist
running egg_info
creating src/tox_travis.egg-info
writing src/tox_travis.egg-info/PKG-INFO
writing dependency_links to src/tox_travis.egg-info/dependency_links.txt
writing entry points to src/tox_travis.egg-info/entry_points.txt
writing requirements to src/tox_travis.egg-info/requires.txt
writing top-level names to src/tox_travis.egg-info/top_level.txt
writing manifest file 'src/tox_travis.egg-info/SOURCES.txt'
reading manifest file 'src/tox_travis.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'src/tox_travis.egg-info/SOURCES.txt'
running check
creating tox-travis-0.11
creating tox-travis-0.11/src
creating tox-travis-0.11/src/tox_travis
creating tox-travis-0.11/src/tox_travis.egg-info
copying files to tox-travis-0.11...
copying HISTORY.rst -> tox-travis-0.11
copying LICENSE -> tox-travis-0.11
copying MANIFEST.in -> tox-travis-0.11
copying README.rst -> tox-travis-0.11
copying setup.cfg -> tox-travis-0.11
copying setup.py -> tox-travis-0.11
copying src/tox_travis/__init__.py -> tox-travis-0.11/src/tox_travis
copying src/tox_travis/after.py -> tox-travis-0.11/src/tox_travis
copying src/tox_travis/envlist.py -> tox-travis-0.11/src/tox_travis
copying src/tox_travis/hacks.py -> tox-travis-0.11/src/tox_travis
copying src/tox_travis/hooks.py -> tox-travis-0.11/src/tox_travis
copying src/tox_travis/utils.py -> tox-travis-0.11/src/tox_travis
copying src/tox_travis.egg-info/PKG-INFO -> tox-travis-0.11/src/tox_travis.egg-info
copying src/tox_travis.egg-info/SOURCES.txt -> tox-travis-0.11/src/tox_travis.egg-info
copying src/tox_travis.egg-info/dependency_links.txt -> tox-travis-0.11/src/tox_travis.egg-info
copying src/tox_travis.egg-info/entry_points.txt -> tox-travis-0.11/src/tox_travis.egg-info
copying src/tox_travis.egg-info/requires.txt -> tox-travis-0.11/src/tox_travis.egg-info
copying src/tox_travis.egg-info/top_level.txt -> tox-travis-0.11/src/tox_travis.egg-info
Writing tox-travis-0.11/setup.cfg
creating dist
Creating tar archive
removing 'tox-travis-0.11' (and everything under it)
```

</details> 

----

If you need all of the files necessary to install and run tests, why not just clone the package at a specific tag? *Note that I'm not familiar with the RPM packaging process.*

**Comment by jayvdb** [comment]
@ryanhiebert, yes it is a bug in the anytemplate setup.py, but it is surprisingly common.
The more typical problem is people forcing installation of tox or flake8 as a 'test' dependency, and openSUSE packaging tools then includes it as a build dependency automatically, and unless someone knows how to patch the sdist, those dependencies stay in the source rpm. e.g. here requiring pytest-flake8

https://build.opensuse.org/request/show/679990

And while it is a bug which can be fixed in the setup.py, people are often reluctant to change their software to suit packaging better.  And even if we do get a patch merged, their sdist doesnt change until their next release, which can be years, or never.

That is the backdrop of my request.

Currently tools *will* add tox-travis as a build dependency of rpms, and there is no package for tox-travis.  So the beginner packager will be confused and/or stop.  Most abandon their effort after they have created packages for a few dependencies - after a few they release they have limited chance of getting all these packages into their distro, so they will need to maintain the spec files themselves.  They only care about the package they started to package, not all the dependencies.  They dont understand which dependencies are necessary and which are not.

In addition, some developers like to have their development environment completely in rpms, so their environment is easily reproducible.  And tox-travis usually ends up in `requirements-dev.txt`, as the best spot to put the version tracking, so it is part of their environment.

Finally, it is not correct to claim tox-travis is only a cloud thing and not a real test dependency.  It is possible for travis-ci to be the *real* test bed, see https://stackoverflow.com/questions/21053657/how-to-run-travis-ci-locally .  Especially https://github.com/SonarSource/local-travis for a real example where a large team is/was using a complete travis stack as the proper way to run the tests locally.

So, I want to create an openSUSE package for tox-travis.

In order to create a package for tox-travis, I need an sdist.  It is that simple.  A wheel is optimised for installation, and not a replacement for an sdist.  A source distribution is where people expect to look to get all the extra bits, like the license file, the changelog, the docs.  While the license is in the wheel, license verification scanners usually only look in sdists.

@rpkilby, absolutely, I can get an sdist from github, and I am doing that in
https://build.opensuse.org/package/view_file/home:jayvdb:moban/python-tox-travis/python-tox-travis.spec?expand=1

It is acceptable, but not ideal as PyPI is the central authority for the versions and contents of the package for each unique identifier, the package name.  It is surprisingly common for the sdist on PyPI to have different source from the contents of a github tag.
As a result, my packaging tools automatically fetch new versions from PyPI.  They don't fetch them from Source in the spec file, because the spec file doesnt indicate how to get the most current release tag, and excluding releases marked pre-releases.  Version numbers jump all over the place.  Also it is very common for people to switch hosting providers, or move repos, and old sdist location has no knowledge of the new releases.  Also common to have hundreds of PyPI packages which are all developed in the same repository.

Of ~thousand PyPI packages I've dealt with over the last six weeks spurt of openSUSE packaging, this is the only one without an sdist in PyPI, except for the many obviously broken packages which are reserving a name for some idea they had but never got off the ground.

**Comment by rpkilby** [comment]
Thanks for the detailed info - that makes a lot of sense since it sounds like you're automating or at least processing these in batch. It would definitely be a huge headache to maintain the list of source repos for hundreds of packages. As you said, it'd be way easier to just rely on PyPI as that central authority.

----

I'm mostly asking for my own reference since the `sdist` call is ultimately @ryanhiebert's, but how do you package different sets of files for `sdist`s and `wheel`s? e.g., if I were to release an `sdist` with all of the package+testing files, how would I also release a `wheel` that excludes the tests? Do you have an example repository I could look at?

**Comment by jayvdb** [comment]
The tests are included in the sdist via `MANIFEST.in`, and they wont go into the wheel.

`python setup.py sdist bdist_wheel` will create the two format ready to upload, but there are tools which wrap up that process, and also do pypi uploading and some do github tagging at the same time.

**Comment by ryanhiebert** [comment]
OK. I don't love it, but I'll go ahead and release with sdists. I'll release an sdist as soon as I confirm that the sdist will contain what we need it to. On that note, running `python setup.py sdist` locally for me created a tarball that _does_ include tests, despite them _not_ being in the `MANIFEST.in`. Any idea why I'm not seeing the same thing as @rpkilby described?

**Comment by rpkilby** [comment]
Weird - maybe an older version of setuptools?



<details>
<summary>Here's the full output of me creating the sdist and checking the files in the resulting tarball.</summary>

```shell
(.venv) Sherpa:tox-travis bagel$ pip list
Package    Version
---------- -------
pip        19.0.3 
setuptools 40.8.0 
wheel      0.33.1 
(.venv) Sherpa:tox-travis bagel$ python setup.py sdist
running sdist
running egg_info
writing src/tox_travis.egg-info/PKG-INFO
writing dependency_links to src/tox_travis.egg-info/dependency_links.txt
writing entry points to src/tox_travis.egg-info/entry_points.txt
writing requirements to src/tox_travis.egg-info/requires.txt
writing top-level names to src/tox_travis.egg-info/top_level.txt
reading manifest file 'src/tox_travis.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'src/tox_travis.egg-info/SOURCES.txt'
running check
creating tox-travis-0.11
creating tox-travis-0.11/src
creating tox-travis-0.11/src/tox_travis
creating tox-travis-0.11/src/tox_travis.egg-info
copying files to tox-travis-0.11...
copying HISTORY.rst -> tox-travis-0.11
copying LICENSE -> tox-travis-0.11
copying MANIFEST.in -> tox-travis-0.11
copying README.rst -> tox-travis-0.11
copying setup.cfg -> tox-travis-0.11
copying setup.py -> tox-travis-0.11
copying src/tox_travis/__init__.py -> tox-travis-0.11/src/tox_travis
copying src/tox_travis/after.py -> tox-travis-0.11/src/tox_travis
copying src/tox_travis/envlist.py -> tox-travis-0.11/src/tox_travis
copying src/tox_travis/hacks.py -> tox-travis-0.11/src/tox_travis
copying src/tox_travis/hooks.py -> tox-travis-0.11/src/tox_travis
copying src/tox_travis/utils.py -> tox-travis-0.11/src/tox_travis
copying src/tox_travis.egg-info/PKG-INFO -> tox-travis-0.11/src/tox_travis.egg-info
copying src/tox_travis.egg-info/SOURCES.txt -> tox-travis-0.11/src/tox_travis.egg-info
copying src/tox_travis.egg-info/dependency_links.txt -> tox-travis-0.11/src/tox_travis.egg-info
copying src/tox_travis.egg-info/entry_points.txt -> tox-travis-0.11/src/tox_travis.egg-info
copying src/tox_travis.egg-info/requires.txt -> tox-travis-0.11/src/tox_travis.egg-info
copying src/tox_travis.egg-info/top_level.txt -> tox-travis-0.11/src/tox_travis.egg-info
Writing tox-travis-0.11/setup.cfg
creating dist
Creating tar archive
removing 'tox-travis-0.11' (and everything under it)
(.venv) Sherpa:tox-travis bagel$ cd dist/
(.venv) Sherpa:dist bagel$ tar -xzf tox-travis-0.11.tar.gz 
(.venv) Sherpa:dist bagel$ tree
.
├── tox-travis-0.11
│   ├── HISTORY.rst
│   ├── LICENSE
│   ├── MANIFEST.in
│   ├── PKG-INFO
│   ├── README.rst
│   ├── setup.cfg
│   ├── setup.py
│   └── src
│       ├── tox_travis
│       │   ├── __init__.py
│       │   ├── after.py
│       │   ├── envlist.py
│       │   ├── hacks.py
│       │   ├── hooks.py
│       │   └── utils.py
│       └── tox_travis.egg-info
│           ├── PKG-INFO
│           ├── SOURCES.txt
│           ├── dependency_links.txt
│           ├── entry_points.txt
│           ├── requires.txt
│           └── top_level.txt
└── tox-travis-0.11.tar.gz

4 directories, 20 files
```

</details>

**Comment by ryanhiebert** [comment]
It's at `40.6.2`, so not all that old. I must be doing something weird. @jayvdb : perhaps you can make a PR for the changes needed for the manifest, and I can take it from there?

**Comment by ryanhiebert** [comment]
sdist is released in 0.12.
